File: Insertion_Execution_Time.py
import random
import time
def insertion_sort_algorithm(array): #Inspiration from Lecture
    steps = 0
    for i in range(1, len(array)):
        key = array[i]
        j = i -1
        steps += 2 
        while j >= 0 and array[j] > key:
            array[j + 1] = array[j]
            j = j - 1
            steps += 2
        array[j + 1] = key
        steps += 1
    return steps

def generate_random_array(size):
    return [random.randint(0, 100) for _ in range(size)]

input_sizes = [100, 1000, 5000, 20000, 50000, 100000]

for size in input_sizes:
    A = generate_random_array(size)
    start_time = time.time()
    steps = insertion_sort_algorithm(A)
    end_time = time.time() - start_time
    print("Size: ", size)
    print("Number of steps: ", steps)
    print("Time taken: ", end_time, "seconds")

File: Insertion_Execution_Time.go
package main

import (
	"fmt"
	"time"
	"math/rand"
)

func insertion_sort(A []int) int {
	steps := 0
	for i := 1; i < len(A); i++ {
		key := A[i]
		j := i - 1
		steps += 2
		for j >= 0 && A[j] > key {
			A[j+1] = A[j]
			j = j -1
			steps += 2
		}
		A[j+1] = key
		steps += 1
	}
	return steps
}

func randomArray(size int) []int {
	A := make([]int, size)
	for i := 0; i < size; i++ {
		A[i] = rand.Intn(100)
	}
	return A
}

func main() {

	input_size := []int{100, 1000, 5000, 20000, 50000, 100000}

	for _, size := range input_size {
		A := randomArray(size)
		start_time := time.Now()
		counting_steps := insertion_sort(A)
		stop_time := time.Since(start_time)
		fmt.Println("Size: ", size)
		fmt.Println("Number of steps: ", counting_steps)
		fmt.Println("Time: ", stop_time)
	}
}